src/gwt/src/org/rstudio/studio/client/workbench/views/files/Files.java

```diff
@@ -607,6 +622,176 @@ void onCopyFilesPaneCurrentDirectory()
       DomUtils.copyCodeToClipboard(currentPath_.getPath());
    }
 
+   @Handler
+   void onOpenFilesInSinglePane()
+   {
+      final ArrayList<FileSystemItem> selectedFiles = view_.getSelectedFiles();
+      final SourceColumnManager mgr = RStudioGinjector.INSTANCE.getSourceColumnManager();
+
+      for (FileSystemItem item : selectedFiles) 
+      {
+         if (!item.isDirectory())
+         {
+            mgr.openFile(item);
+         }
+      }
+
+      view_.selectNone();
+   }
+
+   // the purpose of this function is specifically to pre-screen selected files for the following:
+   //  - the file must NOT be a directory
+   //  - the file must NOT already be open in a source editor 
+   //  - if the file is an RNotebook, then make sure that only ONE .RMarkdown file is opened for editing 
+   private void getUnopenedSelectedFiles(final CommandWithArg<List<FileSystemItem>> onCompleted)
+   {
+      final SourceColumnManager mgr = RStudioGinjector.INSTANCE.getSourceColumnManager();
+      final ArrayList<String> selectedFilePaths = new ArrayList<>(); 
+
+      // Use parallelCommandList to execute these requests as fast as possible in parallel, asynchronously
+      ParallelCommandList commandList = new ParallelCommandList(new Command()
+            {
+               @Override
+               public void execute() 
+               {
+                  final ArrayList<FileSystemItem> selectedFiles = new ArrayList<>(); 
+                  selectedFilePaths.forEach((String path) -> {
+                     selectedFiles.add(FileSystemItem.createFile(path));
+                  });
+                  onCompleted.execute(selectedFiles);
+               }
+            });
+
+
+      final List<FileSystemItem> notebooks = new ArrayList<>();
+
+      // pre-process to make sure there aren't any directories, and that already-open files do not count
+      // towards the column limit
+      for (FileSystemItem item : view_.getSelectedFiles()) 
+      {
+         if (!item.isDirectory() && !mgr.openFileAlreadyOpen(item, null))
+         {
+            TextFileType fileType = fileTypeRegistry_.getTextTypeForFile(item);
+            if (fileType.isRNotebook())
+               notebooks.add(item);
+            else
+               selectedFilePaths.add(item.getPath());
+         }
+      }
+
+      for (FileSystemItem notebook : notebooks)
+      {
+         commandList.addCommand(new SerializedCommand()
+               {
+                  @Override
+                  public void onExecute(final Command continuation)
```

Comment:
Nit here & elsewhere: typically we do not include the additional indentation when the line only contains a bracket. 
```suggestion
         {
            @Override
            public void onExecute(final Command continuation)
```